Module 8: Recurrent Neural Networks- Assignment 

Problem Statement:
Build a sentiment analysis model using Recurrent Neural Networks (RNNs) to
classify movie reviews from the IMDB dataset into positive or negative
sentiments

Dataset:
The dataset comprises 25,000 movie reviews from IMDB, labeled by sentiment
(positive/negative). Reviews have been preprocessed, and each review is
encoded as a sequence of word indices (integers). The words in the dataset are
indexed by overall frequency in the dataset, allowing for quick filtering operations
such as: "only consider the top 10,000 most common words, but eliminate the top
20 most common words"

Tasks to be Performed:
Data Preprocessing:
● Load the IMDB dataset, keeping only the top 10,000 most frequently
occurring words.
● Pad the sequences so that they all have the same length

Model Building:
● Create  sequenctial RNN model using TensorFlow and Keras

● The modle should consist of an Embedding layer , Simple RNN layer and a Dense output layer

● Compile the model, Specefying the appropiate optimizer, loss function and matrices.

Training:
● Train the model on the preprocessed movie reviews, using a batch size of
  128 and validating on 20% of the training data.
● Run the training for 10 epochs

Evaluation:
Evaluate the model on the test set and report the accuracy.
Expected Outcome:
A trained RNN model that can classify movie reviews into positive or negative
sentiments, with an accuracy metric provided at the end of the training process.

Load and Preprocess Data

In [5]:
import tensorflow as tf
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
import numpy as np

print("TensorFlow and Keras modules imported successfully.")

TensorFlow and Keras modules imported successfully.


In [6]:
max_features = 10000
print(f"Loading IMDB dataset with top {max_features} words...")
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

print(f"Original training sequence shape: {x_train.shape}")
print(f"Original testing sequence shape: {x_test.shape}")

# Determine an appropriate maximum sequence length
train_lengths = [len(x) for x in x_train]
max_len = int(np.percentile(train_lengths, 90)) # Use 90th percentile for maxlen

print(f"Calculated maximum sequence length (90th percentile): {max_len}")

# Pad sequences to the chosen maximum length
x_train = pad_sequences(x_train, maxlen=max_len, padding='post', truncating='post')
x_test = pad_sequences(x_test, maxlen=max_len, padding='post', truncating='post')

print(f"Padded training sequence shape: {x_train.shape}")
print(f"Padded testing sequence shape: {x_test.shape}")
print("IMDB dataset loaded and sequences padded successfully.")

Loading IMDB dataset with top 10000 words...
17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 5s 0us/step
Original training sequence shape: (25000,)
Original testing sequence shape: (25000,)
Calculated maximum sequence length (90th percentile): 467
Padded training sequence shape: (25000, 467)
Padded testing sequence shape: (25000, 467)
IMDB dataset loaded and sequences padded successfully.


Build RNN Model

In [8]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense

# Define the RNN model
embedding_dim = 128 # Output dimension for the embedding layer
rnn_units = 128 # Units for the SimpleRNN layer

model = Sequential([
    Embedding(input_dim=max_features, output_dim=embedding_dim, input_length=max_len),
    SimpleRNN(units=rnn_units),
    Dense(1, activation='sigmoid')
])

print("RNN model created successfully.")
model.summary()

RNN model created successfully.


C:\Users\Durga\anaconda3\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn (SimpleRNN)          │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [9]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense

# Define the RNN model
embedding_dim = 128 # Output dimension for the embedding layer
rnn_units = 128 # Units for the SimpleRNN layer

model = Sequential([
    Embedding(input_dim=max_features, output_dim=embedding_dim, input_shape=(max_len,)), # Use input_shape instead of input_length
    SimpleRNN(units=rnn_units),
    Dense(1, activation='sigmoid')
])

print("RNN model created successfully.")
model.summary()

C:\Users\Durga\anaconda3\Lib\site-packages\keras\src\layers\core\embedding.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


RNN model created successfully.


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)         │ (None, 467, 128)       │     1,280,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_1 (SimpleRNN)        │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,313,025 (5.01 MB)

 Trainable params: 1,313,025 (5.01 MB)

 Non-trainable params: 0 (0.00 B)

In [10]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense
from tensorflow.keras import Input

# Define the RNN model
embedding_dim = 128 # Output dimension for the embedding layer
rnn_units = 128 # Units for the SimpleRNN layer

model = Sequential([
    Input(shape=(max_len,)), # Explicitly define Input layer
    Embedding(input_dim=max_features, output_dim=embedding_dim), # No need for input_shape/input_length here
    SimpleRNN(units=rnn_units),
    Dense(1, activation='sigmoid')
])

print("RNN model created successfully.")
model.summary()

RNN model created successfully.


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ (None, 467, 128)       │     1,280,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_2 (SimpleRNN)        │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,313,025 (5.01 MB)

 Trainable params: 1,313,025 (5.01 MB)

 Non-trainable params: 0 (0.00 B)

In [11]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
print("RNN model compiled successfully.")

RNN model compiled successfully.


In [12]:
batch_size = 128
epochs = 10

print(f"Starting model training with batch_size={batch_size}, epochs={epochs}, and validation_split=0.2...")
history = model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_split=0.2)

print("Model training completed. Training history stored.")

Starting model training with batch_size=128, epochs=10, and validation_split=0.2...
Epoch 1/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 95s 572ms/step - accuracy: 0.5070 - loss: 0.6993 - val_accuracy: 0.5062 - val_loss: 0.6975
Epoch 2/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 94s 599ms/step - accuracy: 0.5023 - loss: 0.6986 - val_accuracy: 0.4938 - val_loss: 0.7021
Epoch 3/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 87s 553ms/step - accuracy: 0.5035 - loss: 0.6936 - val_accuracy: 0.4938 - val_loss: 0.6942
Epoch 4/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 136s 515ms/step - accuracy: 0.5054 - loss: 0.6949 - val_accuracy: 0.5034 - val_loss: 0.6952
Epoch 5/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 101s 643ms/step - accuracy: 0.5218 - loss: 0.6898 - val_accuracy: 0.4926 - val_loss: 0.6965
Epoch 6/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 136s 603ms/step - accuracy: 0.5199 - loss: 0.6819 - val_accuracy: 0.4950 - val_loss: 0.6983
Epoch 7/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 137s 574ms/step - accuracy: 0.5379 - loss: 0.6674 - val_accuracy: 0.4926 - val_loss: 0.

In [20]:
print("Evaluating the model on the test set...")
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)

print(f"Test Loss: {loss:.4f}")
print(f"Test Accuracy: {accuracy:.4f}")

Evaluating the model on the test set...
Test Loss: 0.7020
Test Accuracy: 0.5047


Final Task

Summary:

The final accuracy of the sentiment analysis model on the IMDB dataset is 50.00%. This indicates the model performs at a chance level, struggling to effectively classify movie reviews as positive or negative.

Data Analysis Key Findings
*   The IMDB dataset was loaded, retaining the top 10,000 most frequently occurring words.
*   Movie review sequences were padded to a uniform length of 467, determined by the 90th percentile of training sequence lengths.
*   A Recurrent Neural Network (RNN) model was constructed using Keras, comprising an Embedding layer (10000 input dimensions, 128 output dimensions), a SimpleRNN layer (128 units), and a Dense output layer (1 unit with sigmoid activation).
*   The model had a total of 1,313,025 trainable parameters.
*   The model was compiled with the 'adam' optimizer, 'binary\_crossentropy' loss function, and 'accuracy' as the metric.
*   During training over 10 epochs with a batch size of 128 and an 80/20 train/validation split, the training accuracy reached approximately 53.77%, while validation accuracy hovered around 50.62%.
*   Upon evaluation on the test set, the model achieved a test loss of 0.7020 and a test accuracy of 50.47%.

Insights or Next Steps
*   The current SimpleRNN model exhibits poor performance, effectively classifying reviews no better than random chance. This suggests the basic SimpleRNN architecture or its current configuration is insufficient for the complexity of sentiment analysis on the IMDB dataset.
*   To improve model performance, consider using more sophisticated RNN architectures like LSTMs or GRUs, increasing the number of epochs, tuning hyperparameters (e.g., learning rate, embedding dimension, RNN units), or exploring the use of pre-trained word embeddings.
